### ISSUE TITLE
Missing taint: node.kubernetes.io/unreachable:NoExecute when nodes enter NotReady state on new installs of 1.19+

### ISSUE THREAD

**Issue by sculley** [kubernetes/kubernetes#97100]
**What happened**:
I have been testing some _ansible_ roles to deploy _Kubernetes_ using _kubeadm_ on _Ubuntu_ virtual machines, when testing I noticed that pods were not being terminated + re-created on additional nodes when the existing node that the pod lived on went into a _NotReady_ state. The node that goes into a _NotReady_ state does not have the ```node.kubernetes.io/unreachable:NoExecute``` taint applied. Thus preventing the pods to be evicted(terminated) and re-created. (I confirmed this when installing manually(non ansible) too, also I used dockershim/containerd as the cri, tested using flannel/calico as the cni)

This occurs in new installations of _kubernetes_ when using any version greater than _1.18.12_ i.e. _1.19.x_. This has been tested & confirmed with _kubeadm_ running on Ubuntu and kind running on Mac

Strangely I can confirm though, that if _kubeadm/kubelet=1.8.12_ is installed first, then upgraded to _1.19.x_ on Ubuntu using APT that the taint is added to the NotReady node and the pods are evicted(terminated) and re-created on additional nodes. (I haven't confirmed this using kind)

**What you expected to happen**:
I expected that any pods running on a node which has entered the NotReady state to be terminated and re-created on any other available node because the ```node.kubernetes.io/unreachable:NoExecute``` taint is applied.

**How to reproduce it (as minimally and precisely as possible)**:

##### Using kind

```
kind create cluster --config - <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
EOF

kubectl apply -f - <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
spec:
  replicas: 1
  selector:
    matchLabels:
      srv: test
  template:
    metadata:
      labels:
        srv: test
    spec:
      containers:
      - image: nginx
        name: test
        ports:
        - containerPort: 80
        resources: {}
      tolerations:
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        operator: Exists
        tolerationSeconds: 10
      - effect: NoExecute
        key: node.kubernetes.io/not-ready
        operator: Exists
        tolerationSeconds: 10
EOF
```

Working/Tested on k8s 1.18.2 (kind v0.8.1):

```
kubectl get nodes
NAME                 STATUS   ROLES    AGE   VERSION
kind-control-plane   Ready    master   13m   v1.18.2
kind-worker          Ready    <none>   13m   v1.18.2
kind-worker2         Ready    <none>   13m   v1.18.2

kubectl get pods -o wide
NAME                    READY   STATUS    RESTARTS   AGE   IP           NODE          NOMINATED NODE   READINESS GATES
test-67bff98d4b-ch2ng   1/1     Running   0          25s   10.244.2.3   kind-worker   <none>           <none>

docker stop kind-worker
kind-worker

kubectl get nodes
NAME                 STATUS     ROLES    AGE   VERSION
kind-control-plane   Ready      master   23m   v1.18.2
kind-worker          NotReady   <none>   22m   v1.18.2
kind-worker2         Ready      <none>   23m   v1.18.2

kubectl get pods -o wide
NAME                    READY   STATUS        RESTARTS   AGE     IP           NODE           NOMINATED NODE   READINESS GATES
test-67bff98d4b-ch2ng   1/1     Terminating   0          2m38s   10.244.2.3   kind-worker    <none>           <none>
test-67bff98d4b-kqkcv   1/1     Running       0          37s     10.244.3.3   kind-worker2   <none>           <none>

kubectl describe node kind-worker | grep node.kubernetes.io
Taints:             node.kubernetes.io/unreachable:NoExecute
                    node.kubernetes.io/unreachable:NoSchedule
```

Not Working/Tested on k8s 1.19.1 (kind v0.9.0):

```
kubectl get nodes
NAME                 STATUS   ROLES    AGE     VERSION
kind-control-plane   Ready    master   4m37s   v1.19.1
kind-worker          Ready    <none>   4m1s    v1.19.1
kind-worker2         Ready    <none>   4m1s    v1.19.1

kubectl get pods -o wide
NAME                    READY   STATUS    RESTARTS   AGE   IP           NODE           NOMINATED NODE   READINESS GATES
test-5f8dd86d7c-hrlwg   1/1     Running   0          48s   10.244.1.2   kind-worker2   <none>           <none>

docker stop kind-worker2
kind-worker2

kubectl get nodes
NAME                 STATUS     ROLES    AGE     VERSION
kind-control-plane   Ready      master   5m46s   v1.19.1
kind-worker          Ready      <none>   5m10s   v1.19.1
kind-worker2         NotReady   <none>   5m10s   v1.19.1

kubectl get pods -o wide
NAME                    READY   STATUS    RESTARTS   AGE     IP           NODE           NOMINATED NODE   READINESS GATES
test-5f8dd86d7c-hrlwg   1/1     Running   0          3m36s   10.244.1.2   kind-worker2   <none>           <none>

kubectl describe node kind-worker2 | grep node.kubernetes.io
Taints:             node.kubernetes.io/unreachable:NoSchedule
```

##### Using kubeadm

Please see my comment in another [issue](https://github.com/kubernetes/kubernetes/issues/55713#issuecomment-738956787)

**Anything else we need to know?**:

Please see the comment from @adampl confirming the bug [here](https://github.com/kubernetes/kubernetes/issues/55713#issuecomment-739617706) 

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19.x
- Cloud provider or hardware configuration: Bare-metal/Mac
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04/MacOS Catalina
- Kernel (e.g. `uname -a`): ```Linux captain1.k8s.someadmin.com 5.4.0-56-generic #62-Ubuntu SMP Mon Nov 23 19:20:19 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux```
- Install tools: kubeadm/kind
- Network plugin and version (if this is a network-related bug): calico
- Others:

**Comment by pacoxu** [comment]
Cannot reproduce it with 
```
kind v0.9.0 go1.15.5 darwin/amd64
kindest/node:v1.19.1

➜  community git:(feature/image-pull-priority) ✗ kubectl get node
NAME                 STATUS   ROLES    AGE     VERSION
kind-control-plane   Ready    master   8m17s   v1.19.1
kind-worker          Ready    <none>   7m23s   v1.19.1
kind-worker2         Ready    <none>   7m22s   v1.19.1

➜  community git:(feature/image-pull-priority) ✗ docker stop kind-worker2
kind-worker2

➜  community git:(feature/image-pull-priority) ✗ kubectl describe node kind-worker2 | grep node.kubernetes.io

Taints:             node.kubernetes.io/unreachable:NoExecute
                    node.kubernetes.io/unreachable:NoSchedule
```

![image](https://user-images.githubusercontent.com/2010320/101337480-b235b680-38b6-11eb-8ccf-1b5b04080bd1.png)

**Comment by sculley** [comment]
I can still reproduce after deleting and re-creating my cluster using kind just now.

![kind-results](https://user-images.githubusercontent.com/3237705/101339439-54817a00-3876-11eb-88ad-f2da508e4aa6.png)

**Comment by sculley** [comment]
I will run a couple more tests, kind on linux and on my macbook pro to see if can re-produce the same results.

**Comment by sculley** [comment]
I just installed kind and ran on a CentOS 7 box of mine with the same results.

```
uname -a
Linux gertrude 5.6.14-x86_64-linode135 #1 SMP PREEMPT Fri May 22 14:57:20 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

kind --version
kind version 0.9.0

kubectl get nodes
NAME                 STATUS   ROLES    AGE     VERSION
kind-control-plane   Ready    master   3m58s   v1.19.1
kind-worker          Ready    <none>   3m27s   v1.19.1
kind-worker2         Ready    <none>   3m26s   v1.19.1

kubectl get pods -A
NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
kube-system          coredns-f9fd979d6-j5phb                      1/1     Running   0          3m46s
kube-system          coredns-f9fd979d6-zdshm                      1/1     Running   0          3m46s
kube-system          etcd-kind-control-plane                      1/1     Running   0          3m48s
kube-system          kindnet-9vlmk                                1/1     Running   0          3m46s
kube-system          kindnet-nfm9j                                1/1     Running   0          3m31s
kube-system          kindnet-tllcj                                1/1     Running   0          3m30s
kube-system          kube-apiserver-kind-control-plane            1/1     Running   0          3m48s
kube-system          kube-controller-manager-kind-control-plane   1/1     Running   0          3m48s
kube-system          kube-proxy-cgw2k                             1/1     Running   0          3m31s
kube-system          kube-proxy-fp9nr                             1/1     Running   0          3m46s
kube-system          kube-proxy-gt994                             1/1     Running   0          3m30s
kube-system          kube-scheduler-kind-control-plane            1/1     Running   0          3m48s
local-path-storage   local-path-provisioner-78776bfc44-kcfwc      1/1     Running   0          3m46s

docker stop kind-worker2
kind-worker2

kubectl get nodes
NAME                 STATUS     ROLES    AGE     VERSION
kind-control-plane   Ready      master   5m8s    v1.19.1
kind-worker          Ready      <none>   4m37s   v1.19.1
kind-worker2         NotReady   <none>   4m36s   v1.19.1

kubectl describe node kind-worker2 | grep node.kubernetes.io
Taints:             node.kubernetes.io/unreachable:NoSchedule
```

**Comment by adampl** [comment]
FYI: my tests on Kind that confirmed the issue were run on Ubuntu 18.04.5 LTS with Docker 19.03.8

**Comment by pacoxu** [comment]
I can reproduce it on my Ubuntu server, but not on the Mac.

I use kind with docker 19.03.13
```
root@ubuntu-master:~# kubectl get pods -o wide
NAME                    READY   STATUS    RESTARTS   AGE     IP           NODE           NOMINATED NODE   READINESS GATES
test-645b75c986-dlqsw   1/1     Running   0          86s     10.244.1.2   kind-worker2   <none>           <none>
test-645b75c986-ph2lv   1/1     Running   0          6m29s   10.244.2.2   kind-worker    <none>           <none>
root@ubuntu-master:~# kubectl get node
NAME                 STATUS     ROLES    AGE   VERSION
kind-control-plane   Ready      master   28m   v1.19.1
kind-worker          Ready      <none>   28m   v1.19.1
kind-worker2         NotReady   <none>   28m   v1.19.1
root@ubuntu-master:~# kubectl describe node kind-worker2 | grep node.kubernetes.io
Taints:             node.kubernetes.io/unreachable:NoSchedule
root@ubuntu-master:~#
```

**Comment by chenyw1990** [comment]
I also encountered the same problem. This bug has the same reason as #94183 and #96183. and there is an open PR to fixed it #96876

**Comment by sculley** [comment]
One thing I would like to re-mention, when I upgraded a kubeadm install from 1.18.12 to 1.19.4 on Ubuntu, the NoExecute taint is still added but as described here on new installs its not. That seems rather strange?

I would also assume this is quite a serious bug, I imagine most people are not impacted as they user 1.18 and versions provided by cloud providers but this sort of breaks HA in some ways. Glad to see there is a PR to fix though.

**Comment by adampl** [comment]
I never used kubeadm, but I guess it doesn't upgrade existing cluster components automatically.

**Comment by sculley** [comment]
I have upgraded a kubeadm cluster to 1.20.0 from 1.18.12 and I can confirm the correct taints are applied to nodes that are NotReady.

I will be testing spinning up a new cluster using kubeadm with version 1.20.0 and report back with the results.

**Comment by sculley** [comment]
I can confirm on a new install of 1.20.1 the issue still occurs

```shell
kubectl get nodes
NAME                         STATUS     ROLES                  AGE   VERSION
captain1   Ready      control-plane,master   23m   v1.20.1
node1      Ready      <none>                 22m   v1.20.1
node2      Ready      <none>                 22m   v1.20.1
node3      NotReady   <none>                 22m   v1.20.1
```

```shell
kubectl describe node node3 | grep node.kubernetes.io
Taints:             node.kubernetes.io/unreachable:NoSchedule
```

**Comment by guzmonne** [comment]
This PR #96876 has been merged. Does anyone know in which version is available, and if it has indeed fixed the issue?

**Comment by adampl** [comment]
@guzmonne It will appear in 1.21 and there are also backporting PRs for 1.19 (https://github.com/kubernetes/kubernetes/pull/98140) and 1.20 (https://github.com/kubernetes/kubernetes/pull/98168).

**Comment by LuckySB** [comment]
Simple restart API Server help me with this bug. Taint with NoExecute started adding

**Comment by LuckySB** [comment]
1.18.12 has this issue too, can you cherry-pick to 1.18 ?

**Comment by guzmonne** [comment]
Is there anything I can do to help merge (#98168)?

**Comment by adampl** [comment]
@guzmonne Not much, except asking for attention in the PR's discussion section, so that they finally give up :D

**Comment by rainbowjose** [comment]
I can't understand what im doing wrong because a lot people here just keep silence about this problem, so im starting to think that's my fault.

**Comment by ehaiwag** [comment]
I got this issue radomly with kubernetes v1.19.3.
It doesn't help my case with rebooting apiserver.

**Comment by rainbowjose** [comment]
Got same problem, kubernetes v1.20.4

**Comment by guzmonne** [comment]
Until the PR #98168 is merged current versions of 1.20 won't be fixed. It should get done on the following dates.

**Comment by pacoxu** [comment]
> @guzmonne It will appear in 1.21 and there are also backporting PRs for 1.19 (#98140) and 1.20 (#98168).
It will be fixed in 1.20.5 and 1.19.9 if they got merged. The target date is 2021-03-17 according to [detailed-release-history-for-active-branches](https://github.com/kubernetes/sig-release/blob/master/releases/patch-releases.md#detailed-release-history-for-active-branches)

**Comment by hamzakilic** [comment]
I have 3 master HA nodes,
when I shutdown 1 of them, 
kubectl get pods -n kube-system -o wide  
shows all pods Running at down node after 90 minutes
is this the same problem?
![k8bug](https://user-images.githubusercontent.com/11096936/110008455-cdd12180-7d2c-11eb-8fa6-c6c3850e352a.png)

**Comment by adampl** [comment]
@hamzakilic 99% yes.

**Comment by channing342** [comment]
after upgrade to v1.19.9 is normal now 
 
```
System Info:
  Machine ID:                 0865d62057404583a528f06969d7aa8d
  System UUID:                b2450642-5e01-3b93-e39f-9d937b0ee927
  Boot ID:                    ad14c78f-fc72-4095-a5d5-94e19ca68b2f
  Kernel Version:             5.4.0-65-generic
  OS Image:                   Ubuntu 20.04.2 LTS
  Operating System:           linux
  Architecture:               amd64
  Container Runtime Version:  cri-o://1.20.0
  Kubelet Version:            v1.19.9
  Kube-Proxy Version:         v1.19.9
```

![image](https://user-images.githubusercontent.com/4295390/111766680-c9863600-88e0-11eb-98c3-a6aea198cdd7.png)

**Comment by ehashman** [comment]
This was fixed in 1.20.5/1.19.9: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md#changelog-since-v1204

/close
